xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,256 @@
+	inner_product += parameters[i] * fvals[i];
+    }
+
+    return inner_product;
+}
+
+// From Theorem (8) in Cao et al. "Learning to rank: from pairwise approach to listwise approach."
+static prob_distrib_vector
+initializeProbability(vector<FeatureVector> feature_vectors, vector<double> & new_parameters) {
+    LOGCALL_STATIC_VOID(API, "initializeProbability", feature_vectors | new_parameters);
+
+    int list_length = feature_vectors.size();
+
+    // probability distribution, y is ground truth, while z is predict score
+    vector<double> prob_y;
+    vector<double> prob_z;
+    double expsum_y = 0.0;
+    double expsum_z = 0.0;
+
+    for (int i = 0; i < list_length; i++) {
+	expsum_y += exp(feature_vectors[i].get_label());
+	expsum_z += exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()));
+    }
+
+    for (int i = 0; i < list_length; i++) {
+	prob_y.push_back(exp(feature_vectors[i].get_label())/expsum_y);
+	prob_z.push_back(exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()))/expsum_z);
+
+    }
+
+    vector< vector<double> > prob;
+    prob.push_back(prob_y);
+    prob.push_back(prob_z);
+
+    return prob;
+}
+
+// Equation (6) in paper Cao et al. "Learning to rank: from pairwise approach to listwise approach."
+static vector<double>
+calculateGradient(vector<FeatureVector> feature_vectors, prob_distrib_vector prob) {
+    LOGCALL_STATIC_VOID(API, "calculateGradient", feature_vectors | prob_distrib_vector);
+
+    vector<double> gradient(feature_vectors[0].get_fcount(),0);
+    int list_length = feature_vectors.size();
+
+    // Hold ground truth probability distribution
+    vector<double> prob_y(prob[0]);
+    // Hold prediction score probability distribution
+    vector<double> prob_z(prob[1]);
+
+    for (int i = 0; i < list_length; i++) {
+	vector<double> fvals = feature_vectors[i].get_fvals();
+
+	for (int k = 0; k < int(fvals.size()); k++){
+
+	    double first_term = - prob_y[i] * fvals[k];
+	    gradient[k] += first_term;
+
+	    double second_term = prob_z[i] * fvals[k];
+	    gradient[k] += second_term;
+
+	}
+    }
+
+    return gradient;
+}
+
+static void
+updateParameters(vector<double> & new_parameters, vector<double> gradient, double learning_rate) {
+    LOGCALL_STATIC_VOID(API, "updateParameters", new_parameters | gradient | learning_rate);
+	int num = new_parameters.size();
+
+    for (int i = 0; i < num; i++){
+	new_parameters[i] -= gradient[i] * learning_rate;
+    }
+
+}
+
+void
+ListNETRanker::train_model() {
+    LOGCALL_VOID(API, "ListNETRanker::train_model", NO_ARGS);
+
+    std::cout << "ListNET model begin to train..." << endl;
+
+    // get the training data
+    std::vector<Xapian::FeatureVector> fvv = get_traindata();
+    int fvv_len = fvv.size();
+
+    int feature_cnt = -1;
+    if (fvv_len != 0) {
+	feature_cnt = fvv[0].get_fcount();
+    }
+    else {
+	std::cout << "The training data is NULL!" << endl;
+	exit(1); // TODO: Throw exception
+    }
```

Comment:
And it certainly shouldn't call `exit()`.

Resolving this TODO would be good - even just `throw Xapian::InvalidArgumentError("The training data is empty")`. would be an improvement (`NULL` seems the wrong word...)